### ISSUE TITLE
Loss function

### ISSUE THREAD

**Issue by YinTaiWang** [vincentme/GroupRegNet#14]
Hi, I have a question about implementing the loss function. According to the formula on the paper, the code here doesn't seem correct with some missing terms. Additionally, since it's called "normalized" cross-correlation, the loss value should be between 0 and -1, however, sometimes I would get a value like -1.xxx. 
Could you kindly explain why is the implementation of the code and the formula? Thanks a lot in advanced!

```
cross = torch.clamp(IJ_sum - I_sum*J_sum/self.window_volume, min = self.num_stab_const)
I_var = torch.clamp(I2_sum - I_sum**2/self.window_volume, min = self.num_stab_const)
J_var = torch.clamp(J2_sum - J_sum**2/self.window_volume, min = self.num_stab_const)
```

**Comment by vincentme** [comment]
The NCC loss equation in paper is
![图片](https://github.com/vincentme/GroupRegNet/assets/11166542/e7d92b59-5f84-4551-aa99-9cfc31a8da8d)

The NCC loos in code is 
![图片](https://github.com/vincentme/GroupRegNet/assets/11166542/20cd9875-530a-4a3a-9aa9-df1d41063bb2)

I did not see the mismatch

Besides, the NCC is in the range of [-1, 1](https://en.wikipedia.org/wiki/Cross-correlation#Normalization), and the higher means more similar. So the negative is used in loss. The value of -1.xxx you mentioned should be the total loss, not just similarity loss?

**Comment by YinTaiWang** [comment]
I copied the entire code but added the permutation lines because my inputs are in shape (1, n, d, h, w)
![image](https://github.com/vincentme/GroupRegNet/assets/88786569/ac959297-89a7-441b-bc81-135a97c8577e)

I get -1.xxx and it is only the NCC loss
![image](https://github.com/vincentme/GroupRegNet/assets/88786569/45a8a907-de07-4256-9cf5-309d4590a54c)

I re-read the equation and the code. When unpacking the bracket, why does the numerator have only `IJ_sum - I_sum * J_mean` without `- I_mean * J_sum + I_mean * J_mean`? Similarly, some terms seem to be missing from the denominator, or the calculation seems a bit weird. I also reviewed the NCC code from VoxelMorph, and their code seems correct, but my registration result gets messed up after a few epochs. So, I guess I might have some misunderstanding in calculating the equations here.

I've noticed that the equation in your code and the one from VoxelMorph are slightly different, with yours adding a root to the entire equation if I understand it correctly. However, I couldn't understand the missing terms I mentioned above and why the equation here is correct.

Thanks a lot in advance for helping me gain a better understanding of the math here. :)

**Comment by vincentme** [comment]
You are welcome

**numerator have only IJ_sum - I_sum * J_mean**: this is simplification of the computation by expanding the multiplication in the numerator. Something like simplified expression for variance. https://math.stackexchange.com/questions/1040587/need-explanation-how-we-simplified-expression-for-variance

**equation used**: The equation I used should be the standard NCC as you can check [wiki](https://en.wikipedia.org/wiki/Cross-correlation#Normalization) or [matlab doc](https://www.mathworks.com/help/images/ref/normxcorr2.html). 

**ncc_loss less than -1.0**: I can not replicate this behavior. Here is the minimal code to calculate it. The result I got is always around -0.01 which means no correlation. You can try to put your data in. Note: permute data outside the ncc loss to test the ncc function. 

```python
import torch
import torch.nn.functional as F
import torch.nn as nn


class NCC(nn.Module):
    '''
    Calculate local normalized cross-correlation coefficient between tow images. 

    Parameters
    ----------
    dim : int
        Dimension of the input images. 
    windows_size : int
        Side length of the square window to calculate the local NCC. 
    '''
    def __init__(self, dim, windows_size = 11):
        super().__init__()
        assert dim in (2, 3)
        self.dim = dim
        self.num_stab_const = 1e-4 # numerical stability constant
        
        self.windows_size = windows_size
        
        self.pad = windows_size//2
        self.window_volume = windows_size**self.dim
        if self.dim == 2:
            self.conv = F.conv2d
        elif self.dim == 3:
            self.conv = F.conv3d
    
    def forward(self, I, J):
        '''
        Parameters
        ----------
        I and J : (n, 1, h, w) or (n, 1, d, h, w)
            Torch tensor of same shape. The number of image in the first dimension can be different, in which broadcasting will be used. 
        windows_size : int
            Side length of the square window to calculate the local NCC. 
            
        Returns
        -------
        NCC : scalar
            Average local normalized cross-correlation coefficient. 
        '''
        try:
            I_sum = self.conv(I, self.sum_filter, padding = self.pad)
        except:
            self.sum_filter = torch.ones([1, 1] + [self.windows_size, ]*self.dim, dtype = I.dtype, device = I.device)
            I_sum = self.conv(I, self.sum_filter, padding = self.pad)

        J_sum = self.conv(J, self.sum_filter, padding = self.pad) # (n, 1, h, w) or (n, 1, d, h, w)
        I2_sum = self.conv(I*I, self.sum_filter, padding = self.pad)
        J2_sum = self.conv(J*J, self.sum_filter, padding = self.pad)
        IJ_sum = self.conv(I*J, self.sum_filter, padding = self.pad)

        cross = torch.clamp(IJ_sum - I_sum*J_sum/self.window_volume, min = self.num_stab_const)
        I_var = torch.clamp(I2_sum - I_sum**2/self.window_volume, min = self.num_stab_const)
        J_var = torch.clamp(J2_sum - J_sum**2/self.window_volume, min = self.num_stab_const)

        cc = cross/((I_var*J_var)**0.5)
        
        return -torch.mean(cc)

img_size = (4, 1, 64, 128, 256)
num_repeat = 4
ncc = NCC(dim = 3)

ncc_loss_list = torch.zeros(num_repeat, dtype = torch.float32)
for i in range(num_repeat):
    img1 = torch.randn(img_size, dtype = torch.float32)
    img2 = torch.randn(img_size, dtype = torch.float32)
    ncc_loss = ncc(img1, img2)
    ncc_loss_list[i] = ncc_loss
    
print(ncc_loss_list.min())
print(ncc_loss_list.max())
```
